Sources/Kafka/KafkaConsumer.swift

```diff
@@ -78,60 +61,61 @@ public struct KafkaConsumerEvents: Sendable, AsyncSequence {
 
 /// `AsyncSequence` implementation for handling messages received from the Kafka cluster (``KafkaConsumerMessage``).
 public struct KafkaConsumerMessages: Sendable, AsyncSequence {
-    let stateMachine: NIOLockedValueBox<KafkaConsumer.StateMachine>
+    typealias LockedMachine = NIOLockedValueBox<KafkaConsumer.StateMachine>
+
+    let stateMachine: LockedMachine
+    let pollInterval: Duration
 
     public typealias Element = KafkaConsumerMessage
-    typealias BackPressureStrategy = NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark
-    typealias WrappedSequence = NIOThrowingAsyncSequenceProducer<
-        Result<KafkaConsumerMessage, Error>,
-        Error,
-        BackPressureStrategy,
-        KafkaConsumerMessagesDelegate
-    >
-    let wrappedSequence: WrappedSequence
 
     /// `AsynceIteratorProtocol` implementation for handling messages received from the Kafka cluster (``KafkaConsumerMessage``).
     public struct AsyncIterator: AsyncIteratorProtocol {
-        let stateMachine: NIOLockedValueBox<KafkaConsumer.StateMachine>
-        var wrappedIterator: WrappedSequence.AsyncIterator?
+        private let stateMachineHolder: MachineHolder
+        let pollInterval: Duration
 
-        public mutating func next() async throws -> Element? {
-            guard let result = try await self.wrappedIterator?.next() else {
-                self.deallocateIterator()
-                return nil
+        private final class MachineHolder: Sendable { // only for deinit
+            let stateMachine: LockedMachine
+
+            init(stateMachine: LockedMachine) {
+                self.stateMachine = stateMachine
             }
 
-            switch result {
-            case .success(let message):
-                let action = self.stateMachine.withLockedValue { $0.storeOffset() }
+            deinit {
+                self.stateMachine.withLockedValue { $0.finishMessageConsumption() }
+            }
+        }
+
+        init(stateMachine: LockedMachine, pollInterval: Duration) {
+            self.stateMachineHolder = .init(stateMachine: stateMachine)
+            self.pollInterval = pollInterval
+        }
+
+        public func next() async throws -> Element? {
+            // Currently use Task.sleep() if no new messages, should use task executor preference when implemented:
+            // https://github.com/apple/swift-evolution/blob/main/proposals/0417-task-executor-preference.md
```

Comment:
Can we create an issue for this to track it

Reply:
Now we have one https://github.com/swift-server/swift-kafka-client/issues/165 !